File: 0326-power-of-three/0326-power-of-three.go
/*
func isPowerOfThree(n int) bool {
    if n == 1 {
        return true
    }

    if n < 1 || n % 3 != 0 {
        return false
    } 

    return isPowerOfThree(n/3)
}

func isPowerOfThree(n int) bool {
    if n < 1 {
        return false
    }

    for n != 1 {
        if n % 3 != 0 {
            return false
        }
        n /= 3
    }

    return true
}
*/

import (
    "math"
)

func isPowerOfThree(n int) bool {
    if n < 1 {
        return false
    }

    // dividing log(n) by log(3) gives us log_3(n)
    x := math.Round(math.Log(float64(n)) / math.Log(3))

    return n == int(math.Pow(3, x))
}

File: 0326-power-of-three/0326-power-of-three.py
"""
class Solution:
    def isPowerOfThree(self, n: int) -> bool:
        if n < 1:
            return False

        while n != 1:
            if n % 3 != 0:
                return False
    
            n = n // 3    

        return True

class Solution:
    def isPowerOfThree(self, n: int) -> bool:
        if n < 1:
            return False
        elif n == 1:
            return True    

        if n % 3 != 0:
            return False    

        return self.isPowerOfThree(n//3)    
""" 

from math import log

class Solution:
    def isPowerOfThree(self, n: int) -> bool:
        if n < 1:
            return False

        x = round(log(n, 3)) # use round in case imprecise result, e.g. 2.99999999999   
        return n == (3 ** x)